File: sdks/ruby/lib/amlexia/openai.rb
# frozen_string_literal: true

module Amlexia
  module OpenAI
    module_function

    def track_openai_completion(client, model:, status_code:, latency_ms:, usage: {}, cost_usd: nil, endpoint: "/v1/chat/completions", **kwargs)
      event = Cost.enrich_event(
        cost_usd: cost_usd,
        model_name: model,
        provider: "openai",
        tokens_input: usage[:prompt_tokens],
        tokens_output: usage[:completion_tokens],
        total_tokens: usage[:total_tokens]
      )
      client.track(
        endpoint: endpoint,
        method: "POST",
        status_code: status_code,
        latency_ms: latency_ms,
        provider: "openai",
        model_name: model,
        tokens_input: usage[:prompt_tokens],
        tokens_output: usage[:completion_tokens],
        cost_usd: event[:cost_usd],
        **kwargs
      )
    end
  end
end

File: sdks/go/openai.go
package amlexia

type OpenAIUsage struct {
	PromptTokens     int
	CompletionTokens int
	TotalTokens      int
}

func TrackOpenAICompletion(c *Client, model string, statusCode, latencyMs int, usage OpenAIUsage, costUSD *float64) error {
	in := usage.PromptTokens
	out := usage.CompletionTokens
	est := EstimateEventCostUSD(costUSD, model, "openai", in, out, usage.TotalTokens)
	prov := "openai"
	endpoint := "/v1/chat/completions"
	method := "POST"
	return c.Track(Event{
		Endpoint:   endpoint,
		Method:     method,
		StatusCode: statusCode,
		LatencyMs:  latencyMs,
		CostUSD:    est,
		Provider:   &prov,
		ModelName:  &model,
		TokensIn:   &in,
		TokensOut:  &out,
	})
}
